File: r153/src/eqwit153.c
/* Round 153 -- INDEPENDENT enumeration of the equality witnesses.
 *
 * WHAT IS ENUMERATED.  For a cell K = (b, d, a, bb, e, h) and a target T this
 * lists EVERY walk of the round-152 transition system that starts at 123456,
 * respects all six budgets, has exactly T ports and ends with deficit <= d.
 *
 * Round 148 recorded a witness only when the deficit was EXACTLY d.  This
 * enumerator is deliberately more inclusive -- deficit <= d -- so that a
 * realisation with a smaller deficit cannot slip through unlisted.  The
 * deficit of each witness is reported, so the two conventions can be compared.
 *
 * WHY STOPPING AT DEPTH T IS COMPLETE.  Round 152 certified cap(K) = T, so no
 * walk has more than T ports at deficit <= d; a longer walk can therefore never
 * become a witness.  That certified value is the ONLY round-152 input used
 * here (Phase 9: do not recompute the capacity tables).
 *
 * PRUNING.  Two routes:
 *   --no-table   only the proved analytic reach bound (P2) and the round-150
 *                feasibility lemma.  Nothing else.
 *   --table F    additionally the (P1) monotone bound, taken from a list of
 *                ROUND-152 CERTIFIED values (same format as the checker's
 *                prune list: "<b> <d> <a> <bb> <e> <h> <cap>" per line).
 * The two routes must produce the same witness set.
 *
 * build: cc -O2 -o eqwit153.exe eqwit153.c
 * run:   ./eqwit153.exe <b> <d> <a> <bb> <e> <h> <target> <out.jsonl>
 *                       [--table <prune.txt> | --no-table] [node_cap]
 */
#define CHECKER152_NO_MAIN
#include "checker152.c"

static FILE *OUT;
static uint64_t WITCOUNT;
static int ETARGET, USE_TABLE;

static void erec(int cur, int corb, int ports, int tok, int au, int bu,
                 int eu, int hu) {
    ++NODES;
    if (NODECAP && NODES > NODECAP) { CAPPED = 1; return; }
    if (ports == ETARGET) {
        if (deficit <= CD) {
            ++WITCOUNT;
            fprintf(OUT, "{\"deficit\":%d,\"ports\":[", deficit);
            for (int i = 0; i < ports; ++i)
                fprintf(OUT, "%s%d", i ? "," : "", TRAIL[i]);
            fputs("]}\n", OUT);
            if (ferror(OUT)) { fprintf(stderr, "witness write\n"); exit(3); }
        }
        return;                       /* cap(K) = T: nothing longer can qualify */
    }
    if (!feas(tok, corb)) return;
    int reach = ports + (NH - hexcount) + (CA - au) + (CBB - bu) + (CE - eu);
    if (USE_TABLE) {
        int r2 = ports + ubound(tok, CD - deficit + 4 + 5 * tok, CA - au,
                                CBB - bu, CE - eu, CH - hu) - 1;
        if (r2 < reach) reach = r2;
    }
    if (reach < ETARGET) return;

    /* clean E */
    int t = FREE_T[cur];
    {
        int q = ORBID[t];
        if (q == corb && !(phm[q] >> PHASE[t] & 1)) {
            int newhex = !hexu[HEXID[t]];
            if (newhex || CE > eu) {
                int spend = newhex ? 0 : 1;      /* clean E is not dirty */
                unsigned char old = phm[q];
                phm[q] = (unsigned char)(old | (1u << PHASE[t]));
                deficit -= 1;
                if (newhex) { hexu[HEXID[t]] = 1; ++hexcount; }
                TRAIL[ports] = t;
                erec(t, q, ports + 1, tok, au, bu, eu + spend, hu);
                if (newhex) { hexu[HEXID[t]] = 0; --hexcount; }
                deficit += 1;
                phm[q] = old;
            }
        }
    }
    /* dirty A, dirty B, five paid, heavy */
    for (int kind = 0; kind < 4; ++kind) {
        int lo = 0, hi = 0;
        if (kind == 0) { if (au >= CA) continue; hi = 1; }
        if (kind == 1) { if (bu >= CBB) continue; hi = 1; }
        if (kind == 2) hi = 5;
        if (kind == 3) hi = NHV[cur];
        for (int i = lo; i < hi; ++i) {
            int tt, cost = 0, isdirty;
            if (kind == 0) { tt = DA_T[cur]; isdirty = 1; }
            else if (kind == 1) { tt = DB_T[cur]; isdirty = 2; }
            else if (kind == 2) { tt = PAID_T[cur][i]; isdirty = 0; }
            else { tt = HV_T[cur][i]; cost = HV_C[cur][i]; isdirty = 0;
                   if (hu + cost > CH) continue; }
            int q = ORBID[tt];
            if (phm[q] >> PHASE[tt] & 1) continue;
            int newhex = !hexu[HEXID[tt]], spend = 0;
            if (!newhex && isdirty == 0) { if (eu >= CE) continue; spend = 1; }
            int fresh = (phm[q] == 0);
            int tc = fresh ? 0 : 1;
            if (tc > tok) continue;
            unsigned char old = phm[q];
            phm[q] = (unsigned char)(old | (1u << PHASE[tt]));
            if (fresh) { opened[nopened++] = q; deficit += 4; } else deficit -= 1;
            if (newhex) { hexu[HEXID[tt]] = 1; ++hexcount; }
            TRAIL[ports] = tt;
            erec(tt, q, ports + 1, tok - tc, au + (isdirty == 1),
                 bu + (isdirty == 2), eu + spend, hu + cost);
            if (newhex) { hexu[HEXID[tt]] = 0; --hexcount; }
            if (fresh) { --nopened; deficit -= 4; } else deficit += 1;
            phm[q] = old;
        }
    }
}

int main(int argc, char **argv) {
    if (argc < 9) {
        fprintf(stderr, "usage: %s <b> <d> <a> <bb> <e> <h> <target> <out.jsonl>"
                        " [--table <prune.txt>|--no-table] [node_cap]\n", argv[0]);
        return 3;
    }
    CB = atoi(argv[1]); CD = atoi(argv[2]); CA = atoi(argv[3]);
    CBB = atoi(argv[4]); CE = atoi(argv[5]); CH = atoi(argv[6]);
    ETARGET = atoi(argv[7]);
    common_setup();
    int ai = 9;
    const char *prune = NULL;
    if (argc > ai && !strcmp(argv[ai], "--table")) { prune = argv[ai + 1]; ai += 2; }
    else if (argc > ai && !strcmp(argv[ai], "--no-table")) { ++ai; }
    NODECAP = argc > ai ? strtoull(argv[ai], NULL, 10) : 0;
    if (prune) {
        read_prune(prune);
        int self[6] = {CB, CD, CA, CBB, CE, CH};
        (void)self;
        ub_rebuild_excluding((int[6]){-1, -1, -1, -1, -1, -1});  /* keep all */
        USE_TABLE = 1;
    }
    OUT = fopen(argv[8], "w");
    if (!OUT) { fprintf(stderr, "cannot write %s\n", argv[8]); return 3; }
    NODES = 0; CAPPED = 0; WITCOUNT = 0;
    memset(phm, 0, sizeof phm); memset(hexu, 0, sizeof hexu);
    nopened = 0; hexcount = 1; hexu[HEXID[0]] = 1;
    phm[ORBID[0]] = (unsigned char)(1u << PHASE[0]);
    opened[nopened++] = ORBID[0]; deficit = 4;
    TRAIL[0] = 0;
    clock_t t0 = clock();
    erec(0, ORBID[0], 1, CB, 0, 0, 0, 0);
    fclose(OUT);
    printf("{\"enumerator\":\"EQ153\",\"cell\":\"%d|%d|%d|%d|%d|%d\","
           "\"target\":%d,\"route\":\"%s\",\"witnesses\":%llu,\"nodes\":%llu,"
           "\"capped\":%s,\"seconds\":%.2f,\"file\":\"%s\"}\n",
           CB, CD, CA, CBB, CE, CH, ETARGET, prune ? "table" : "no-table",
           (unsigned long long)WITCOUNT, (unsigned long long)NODES,
           CAPPED ? "true" : "false",
           (double)(clock() - t0) / CLOCKS_PER_SEC, argv[8]);
    return CAPPED ? 1 : 0;
}

File: r153/src/eqwit153.py
#!/usr/bin/env python3
"""Round 153 -- a THIRD equality-witness enumerator, and the witness checker.

Three things live here.

  enumerate   an independent Python enumeration of every walk with exactly T
              ports and final deficit <= d.  It uses the catalogue that
              r152/src/checker152.py builds from string algebra, visits the
              moves in the REVERSE of the C enumerator's order, and prunes with
              nothing but the two proved facts (the analytic reach bound and
              the round-150 feasibility lemma).  A different traversal order
              gives different node counts; the SET it produces must be the same.

  check       replay a stored witness move by move: every step a catalogue
              joint, no phase reused, clean E inside its orbit, every budget
              respected, exactly T ports, final deficit <= d, and the deficit
              identity.  A witness that does not replay is not a witness.

  canonical   the least relabelling of the six letters (left S6, 720 elements),
              which is the only symmetry used anywhere in this round.  Left S6
              is a bijection of the 720 words that commutes with sigma, tau,
              end and the overlap length, so it maps walks to walks -- that is
              round 149's Theorem 2.4, checked exhaustively there over all
              518,400 (group element, word) pairs and re-checked here.

usage: eqwit153.py enumerate <b> <d> <a> <bb> <e> <h> <target> <out.jsonl>
       eqwit153.py check <b> <d> <a> <bb> <e> <h> <target> <file.jsonl> [...]
"""
from __future__ import annotations
import hashlib, itertools, json, sys, time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent.parent
                       / "r152" / "src"))
from checker152 import (PERMS, IDX, HEX, ORB, PHASE, MOVES, FREE, DA, DB,   # noqa
                        PAID, HEAVY, N, NHEX, _sigma, _tau, _end, _gap)

HEXCAP = 120


# ----------------------------------------------------------------- symmetry
def relabel_map(g):
    m = dict(zip("123456", g))
    return [IDX["".join(m[ch] for ch in w)] for w in PERMS]


_RL = None


def relabellings():
    global _RL
    if _RL is None:
        _RL = [relabel_map(g) for g in itertools.permutations("123456")]
    return _RL


def canonical(ports):
    best = None
    for r in relabellings():
        t = tuple(r[v] for v in ports)
        if best is None or t < best:
            best = t
    return best


def equivariance_check():
    """Left S6 commutes with the catalogue: all 720 x 720 pairs."""
    bad = 0
    for r in relabellings():
        for v in range(N):
            if FREE[r[v]] != r[FREE[v]] or DA[r[v]] != r[DA[v]] \
               or DB[r[v]] != r[DB[v]]:
                bad += 1
                break
            if sorted(PAID[r[v]]) != sorted(r[t] for t in PAID[v]):
                bad += 1
                break
            if sorted((r[t], c) for t, c in HEAVY[v]) != sorted(HEAVY[r[v]]):
                bad += 1
                break
    return dict(group=len(relabellings()), words=N,
                pairs=len(relabellings()) * N, violations=bad, ok=bad == 0)


# ---------------------------------------------------------------- the check
def check_witness(cell, target, ports):
    b, dmax, amax, bmax, emax, hmax = cell
    if len(ports) != target:
        return False, f"{len(ports)} ports, target {target}"
    if not ports or ports[0] != IDX["123456"]:
        return False, "does not start at 123456"
    if len(set(ports)) != len(ports):
        return False, "a port repeats"
    mv = {}
    for v in range(N):
        for t, kind, cost in MOVES[v]:
            mv.setdefault((v, t), (kind, cost))
    phm = {ORB[ports[0]]: {PHASE[ports[0]]}}
    hexu = {HEX[ports[0]]}
    deficit, tok, au, bu, eu, hu = 4, 0, 0, 0, 0, 0
    corb = ORB[ports[0]]
    for i in range(len(ports) - 1):
        u, t = ports[i], ports[i + 1]
        if (u, t) not in mv:
            return False, f"step {i} is not a catalogue joint"
        kind, cost = mv[(u, t)]
        q = ORB[t]
        if PHASE[t] in phm.get(q, ()):
            return False, f"step {i} reuses a phase"
        if kind == "E" and q != corb:
            return False, f"step {i}: clean E left its orbit"
        fresh = q not in phm
        newhex = HEX[t] not in hexu
        if not newhex and kind not in ("A", "B"):
            eu += 1
        au += kind == "A"
        bu += kind == "B"
        hu += cost if kind == "H" else 0
        tok += 0 if (kind == "E" or fresh) else 1
        phm.setdefault(q, set()).add(PHASE[t])
        hexu.add(HEX[t])
        deficit += 4 if fresh else -1
        corb = q
    if tok > b or deficit > dmax or au > amax or bu > bmax or eu > emax \
       or hu > hmax:
        return False, (f"budget: tok={tok}/{b} def={deficit}/{dmax} "
                       f"a={au}/{amax} bb={bu}/{bmax} e={eu}/{emax} "
                       f"h={hu}/{hmax}")
    if deficit != 5 * len(phm) - len(ports):
        return False, "deficit identity violated"
    return True, dict(ports=len(ports), deficit=deficit, orbits=len(phm),
                      hexagons=len(hexu), hex_simple=len(hexu) == len(ports),
                      tok=tok, a=au, bb=bu, e=eu, h=hu)


# ------------------------------------------------------------ the traversal
def enumerate_witnesses(cell, target, node_cap=0):
    b, dmax, amax, bmax, emax, hmax = cell
    phm = {ORB[0]: {PHASE[0]}}
    hexc = {HEX[0]: 1}
    path = [0]
    out, st = [], dict(nodes=0)

    def feas(corb, tok):
        us = sorted((5 - len(ph) for q, ph in phm.items() if q != corb),
                    reverse=True)
        return sum(us[tok:]) <= dmax

    def rec(v, corb, ports, tok, au, bu, eu, hu, deficit):
        st["nodes"] += 1
        if node_cap and st["nodes"] > node_cap:
            raise TimeoutError
        if ports == target:
            if deficit <= dmax:
                out.append(dict(deficit=deficit, ports=list(path)))
            return
        if not feas(corb, tok):
            return
        if ports + (NHEX - len(hexc)) + (amax - au) + (bmax - bu) \
           + (emax - eu) < target:
            return
        for t, kind, cost in reversed(MOVES[v]):      # the reverse order
            q = ORB[t]
            ph = phm.get(q)
            if ph is not None and PHASE[t] in ph:
                continue
            if kind == "E" and q != corb:
                continue
            if kind == "A" and au >= amax:
                continue
            if kind == "B" and bu >= bmax:
                continue
            if kind == "H" and hu + cost > hmax:
                continue
            newhex = HEX[t] not in hexc
            se = 0
            if not newhex and kind not in ("A", "B"):
                if eu >= emax:
                    continue
                se = 1
            fresh = ph is None
            c = 0 if (kind == "E" or fresh) else 1
            if c > tok:
                continue
            if fresh:
                phm[q] = {PHASE[t]}
            else:
                ph.add(PHASE[t])
            hexc[HEX[t]] = hexc.get(HEX[t], 0) + 1
            path.append(t)
            rec(t, q, ports + 1, tok - c, au + (kind == "A"),
                bu + (kind == "B"), eu + se,
                hu + (cost if kind == "H" else 0),
                deficit + (4 if fresh else -1))
            path.pop()
            hexc[HEX[t]] -= 1
            if not hexc[HEX[t]]:
                del hexc[HEX[t]]
            if fresh:
                del phm[q]
            else:
                ph.discard(PHASE[t])

    sys.setrecursionlimit(10000)
    rec(0, ORB[0], 1, b, 0, 0, 0, 0, 4)
    return out, st["nodes"]


def load(p):
    return [json.loads(x) for x in Path(p).read_text().splitlines() if x.strip()]


def fingerprint(witnesses):
    cs = sorted(canonical(w["ports"]) for w in witnesses)
    classes = sorted(set(cs))
    return dict(raw=len(witnesses), classes=len(classes),
                canonical_sha256=hashlib.sha256(
                    json.dumps([list(c) for c in classes]).encode()).hexdigest(),
                raw_sha256=hashlib.sha256(json.dumps(
                    sorted(tuple(w["ports"]) for w in witnesses)).encode()
                ).hexdigest(),
                deficits=sorted({w["deficit"] for w in witnesses}),
                classes_list=[list(c) for c in classes])


def main(argv):
    mode = argv[0]
    cell = tuple(int(x) for x in argv[1:7])
    target = int(argv[7])
    if mode == "enumerate":
        t0 = time.time()
        wits, nodes = enumerate_witnesses(cell, target)
        Path(argv[8]).write_text(
            "".join(json.dumps(w) + "\n" for w in wits))
        print(json.dumps(dict(enumerator="PY153", cell="|".join(map(str, cell)),
                              target=target, route="no-table",
                              witnesses=len(wits), nodes=nodes,
                              seconds=round(time.time() - t0, 2),
                              file=argv[8])))
        return 0
    if mode == "check":
        ok = True
        for p in argv[8:]:
            ws = load(p)
            for w in ws:
                good, info = check_witness(cell, target, w["ports"])
                if not good:
                    ok = False
                    print(f"  BAD {p}: {info}")
                elif info["deficit"] != w["deficit"]:
                    ok = False
                    print(f"  BAD {p}: stored deficit {w['deficit']} != "
                          f"{info['deficit']}")
            fp = fingerprint(ws)
            print(f"  {p}: raw={fp['raw']} classes={fp['classes']} "
                  f"deficits={fp['deficits']} "
                  f"canon={fp['canonical_sha256'][:16]}")
        return 0 if ok else 1
    raise SystemExit(f"unknown mode {mode}")


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))
